Initialize Builtins at Native Image build time

hubertp opened the pull request:
### Pull Request Description

Moved loading of Builtin Types and Methods to a static initializer. That way the information is available at Native Image build time and one does not have to update a corresponding entry in `reflect-config` which would be a real pain when we start using it in anger.
Fixes https://www.pivotaltracker.com/story/show/183374932

### Important notice

For the inliner to do its job, we have to limit call chains. Therefore rather than making `clazz.getConstructor().newInstance()` calls we should do `clazz.getConstructor()`. The latter will get constant folded and won't crash when running NI, the former won't.

hubertp committed:
Initialize Builtins at build time

Moved loading of Builtin Types and Methods to a static initializer.
That way the information is available at Native Image build time and one
does not have to update a corresponding entry in `reflect-config` which
would be a real pain when we start using it in anger.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java:
```diff
@@ -48,6 +48,14 @@
 
 /** Container class for static predefined atoms, methods, and their containing scope. */
 public class Builtins {
+  private static Map<String, Method> builtinMethods;
```
I'd made the fields `final`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java:
```diff
@@ -149,13 +159,12 @@ private void registerBuiltinMethods(ModuleScope scope, Language language) {
               Optional<Function> fun;
               try {
-                Method meth = clazz.getMethod("makeFunction", Language.class);
                 fun = Optional.ofNullable((Function) meth.invoke(null, language));
```
Does this really work without any configuration? Is `native-image` clever enough to register existing `Method`instances in the heap for reflection?

hubertp replied:
Yeap, it does work.

hubertp replied:
None that Graal treats `getMethod` a bit specially. Full list is available at https://www.graalvm.org/22.0/reference-manual/native-image/Reflection/

JaroslavTulach commented on engine/runner-native/src/main/resources/META-INF/native-image/org/enso/runner/reflect-config.json:
```diff
@@ -42,1272 +42,6 @@
-    "methods":[{"name":"<init>","parameterTypes":[] }]}
-,
-  {
-    "name":"org.enso.interpreter.node.expression.builtin.number.bigInteger.AbsMethodGen",
```
Nice removals.

JaroslavTulach reviewed (approved):
I am pleasantly surprised that one can `Method.invoke` without specifying that in the `reflect-config.json`!

Akirathan commented:
Suggested edit:
```diff
diff --git a/.metals/metals.lock.db b/.metals/metals.lock.db
new file mode 100644
index 000000000..3ef31c8b7
--- /dev/null
+++ b/.metals/metals.lock.db
@@ -0,0 +1,6 @@
+#FileLock
+#Fri Oct 21 13:23:00 CEST 2022
+server=localhost\:40373
+hostName=localhost
+method=file
+id=183fa4837462a56c750b828d490dd0256184acc725e
diff --git a/.metals/metals.mv.db b/.metals/metals.mv.db
new file mode 100644
index 000000000..70ae00b95
Binary files /dev/null and b/.metals/metals.mv.db differ
diff --git a/.metals/workspace-symbol.md b/.metals/workspace-symbol.md
new file mode 100644
index 000000000..5329031f3
--- /dev/null
+++ b/.metals/workspace-symbol.md
@@ -0,0 +1 @@
+Add ';' to search library dependencies
\ No newline at end of file
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java
index 44a15e35a..dcdca6f10 100644
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java
@@ -313,8 +313,6 @@ public class Builtins {
    * single builtin method per row. The format of the row is as follows: <Fully qualified name of
    * the builtin method>:<Class name of the builtin method representing it>
    *
-   * @param scope Builtins scope
-   * @param classes a map of (already loaded) builtin methods
    * @return A map of builtin method nodes per builtin type name
    */
   private static Map<String, Method> readBuiltinMethodsMethods() {

```

Akirathan commented:
Suggested edit:
```diff
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java
index 44a15e35a..dcdca6f10 100644
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java
@@ -313,8 +313,6 @@ public class Builtins {
    * single builtin method per row. The format of the row is as follows: <Fully qualified name of
    * the builtin method>:<Class name of the builtin method representing it>
    *
-   * @param scope Builtins scope
-   * @param classes a map of (already loaded) builtin methods
    * @return A map of builtin method nodes per builtin type name
    */
   private static Map<String, Method> readBuiltinMethodsMethods() {

```

hubertp committed:
changelog

hubertp commented:
This currently breaks because tests call Builtins multiple times and some state is modified during initialization. Resetting the state of types and scope would be recommended but it opens its own can of worms that I don't want to do.

On the other hand, we could statically just initialize the class and then create new instances in Builtins constructor. Unfortunately calls are then not inlined, we get `NoSuchMethodException` which looks like a limitation of https://github.com/oracle/graal/issues/2500. I will see if I can refactor it a bit so that it is more inliner friendly.

hubertp commented:
> On the other hand, we could statically just initialize the class and then create new instances in Builtins constructor. Unfortunately calls are then not inlined, we get `NoSuchMethodException` which looks like a limitation of [oracle/graal#2500](https://github.com/oracle/graal/issues/2500). I will see if I can refactor it a bit so that it is more inliner friendly.

Yeap, that was it. As soon as I moved `.newInstance()` out of `clazz.getConstructor().newInstance()` in the static initializer to the `Builtins` constructor, it had no problem with inlining the call. I wish there were more diagnostics telling us that we are hitting some limits because sometimes it feels like walking through a fog with Graal.

hubertp committed:
Always create new builtin instances

Turns out one cannot re-use builtins that are initialized in the static
context because some state is not reset between tests. Trying to enforce
that will also open a can of worms so we decided against that.

However, while trying to instantiate builtins in the constructor
we would no longer inling constructor calls, resulting in
`NoSuchMethodException` once the native image was generated. This turned
out to be a limitation related to https://github.com/oracle/graal/issues/2500
so that we cannot make calls like `clazz.getContructor().newInstance()`.
`clazz.getConstructor()` in the static initializer and creating new
instances in Builtins constructor was however OK for the inliner and it
was able to procede with constant folding.

hubertp committed:
nit

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/builtins-static-init-183374932
